### ISSUE TITLE
how to customize seed

### ISSUE THREAD

**Issue by dodoyoon** [rohanpadhye/JQF#131]
Hello, I am new to using JQF and I am trying to find ways to customize the seed of the input. 

I did read the issue that allows to resume fuzzing with previously created corpus. 
But is there an option that I can set the seed of the input?

**Comment by rohanpadhye** [comment]
Hello! Using a seed and resuming fuzzing has basically the same usage. Run `mvn jqf:fuzz -Din=<seeds_dir> ...`. 

However, typically it is hard to come up with meaningful seeds for arbitrary generators, so this is rarely used beyond resuming a previous fuzzing session. For example, it is used in [Bonsai Fuzzing](https://arxiv.org/pdf/2103.04388.pdf) to take inputs that were produced by a different fuzzer that used the same generator.

**Comment by dodoyoon** [comment]
Thank you for the response! 

I have another quick question. 
I would like to give specific constant as a seed, like "# comment\na,b,c\n1,2,3\nx,y,z" for String. 
Since corpus generated by jqf:fuzz is not in human-readable format, I was wondering this was possible.

**Comment by rohanpadhye** [comment]
It is not in human readable format because in general the inputs can be
more than one parameter of arbitrary type. The files simply store the
choices made in the generator.

If your test driver takes in a single input that can be serialized to file,
then try using exactly one parameter of type InputStream in the @Fuzz
annotated method. In this way, your test will use the InputStreamGenerator
which simply emits the same bytes as in the backing source. Then, the seed
inputs and fuzzer saved files will all be "human readable" ; that is, the
values in the file so be the same as the value your input stream gets in
the test method.

Then you can use seed inputs as you suggest. Also check out this generator
for String type inputs thag dosa the same thing by using an underlying
InputStream:
https://github.com/rohanpadhye/JQF/blob/master/examples/src/main/java/edu/berkeley/cs/jqf/examples/common/ArbitraryLengthStringGenerator.java
.

**Comment by dodoyoon** [comment]
Thank you, again. 

I tried running the method 'testWithInputStream(InputStream in)', but when I have a look at the corpus, it still seems to be in non-"human readable" format. 
https://github.com/rohanpadhye/JQF/blob/c5feaffc09ff85fbbffdce818ff0179b0a4883e2/examples/src/test/java/edu/berkeley/cs/jqf/examples/rhino/CompilerTest.java 
But I will try to figure it out. 

Anyhow, just to be sure. Is there no way to provide seeds like below? (perhaps as a text file or other type of file)
![image](https://user-images.githubusercontent.com/36918177/113504414-eecaa380-9572-11eb-8f62-4b4f42aca4f8.png)

**Comment by rohanpadhye** [comment]
Interesting. I see what you mean. Let me take a look to see if it's a bug. Might take a couple of days.

**Comment by rohanpadhye** [comment]
I've pushed a patch that might help with this a bit. The main issue with the Rhino test is that it tries to read the entire input stream until EOF before parsing, which is not natural for the way Zest is implemented. In Zest, inputs are usually represented as an infinite stream of choices for the generator, and they can be dynamically extended on-demand. This means that if your input is of type `InputStream` and the program under test doesn't parse the input lazily, it will just keep reading input bytes forever and then probably crash in a non-reproducible way.

Commit 940c7c6  adds support for option `-DfixedSize`, which will force Zest to mutate seed inputs in-place. In the latest version (install the maven plugin SNAPSHOT by running `mvn install` after cloning the JQF repository), try running the following:
```
mvn jqf:fuzz -Dclass=edu.berkeley.cs.jqf.examples.rhino.CompilerTest -Dmethod=testWithInputStream -Din=seeds/ -DfixedSize
```

I tried this with a single seed input having the string `console.log(1)` and it was immediately recognized as a valid input:

![image](https://user-images.githubusercontent.com/4233266/114228113-ee9c2080-9943-11eb-9e23-43e5a8e2f2ba.png)


Note that this is not an application that Zest was designed for, and it has several limitations (e.g. you can't provide dictionaries and statically change the size of mutated inputs). If you really want to fuzz Java programs with files rather than Java objects, maybe try using the [JQF-AFL plugin](https://github.com/rohanpadhye/jqf/wiki/Fuzzing-with-AFL) instead of Zest.

**Comment by dodoyoon** [comment]
This is a great help. Thank you very much!

**Comment by rohanpadhye** [comment]
The updates are now available in JQF release 1.7. Thanks for the issue report!
